flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/internal/converter/AbstractJdbcRowConverter.java

```diff
@@ -114,49 +119,54 @@ protected JdbcDeserializationConverter createNullableInternalConverter(LogicalTy
 	}
 
 	protected JdbcDeserializationConverter wrapIntoNullableInternalConverter(JdbcDeserializationConverter jdbcDeserializationConverter) {
-		return v -> {
-			if (v == null) {
+		return (val, sqlType, unsigned) -> {
+			if (val == null) {
 				return null;
 			} else {
-				return jdbcDeserializationConverter.deserialize(v);
+				return jdbcDeserializationConverter.deserialize(val, sqlType, unsigned);
 			}
 		};
 	}
 
 	protected JdbcDeserializationConverter createInternalConverter(LogicalType type) {
 		switch (type.getTypeRoot()) {
 			case NULL:
-				return v -> null;
+				return (val, sqlType, unsigned) -> null;
 			case BOOLEAN:
-			case TINYINT:
 			case FLOAT:
 			case DOUBLE:
-			case INTEGER:
 			case INTERVAL_YEAR_MONTH:
-			case BIGINT:
 			case INTERVAL_DAY_TIME:
-				return v -> v;
+				return (val, sqlType, unsigned) -> val;
+			case TINYINT:
+				return (val, sqlType, unsigned) -> ((Integer) val).byteValue();
 			case SMALLINT:
 				// Converter for small type that casts value to int and then return short value, since
 				// JDBC 1.0 use int type for small values.
-				return v -> (Integer.valueOf(v.toString())).shortValue();
+				return (val, sqlType, unsigned) -> (Integer.valueOf(val.toString())).shortValue();
+			case INTEGER:
+				return (val, sqlType, unsigned) -> sqlType == Types.SMALLINT && unsigned ? Integer.valueOf(val.toString()) : val;
+			case BIGINT:
+				return (val, sqlType, unsigned) -> sqlType == Types.INTEGER && unsigned ? Long.valueOf(val.toString()) : val;
+			case DECIMAL:
+				final int precision = ((DecimalType) type).getPrecision();
+				final int scale = ((DecimalType) type).getScale();
+				// using decimal(20, 0) to map bigint unsigned
+				return (val, sqlType, unsigned) -> sqlType == Types.BIGINT && unsigned ?
+					DecimalData.fromBigDecimal(new BigDecimal((BigInteger) val, 0), 20, 0) : DecimalData.fromBigDecimal((BigDecimal) val, precision, scale);
```

Comment:
I suggest to use the precision and scale from the `DecimalType` instead of hard code `20, 0`, otherwise, if users use a `30, 0` in DDL, the value will be corrupt.